src/allocator.rs

```diff
@@ -4,17 +4,26 @@
  *
  */
 
-pub enum SExp<'a, T> {
-    Atom(&'a [u8]),
+pub enum SExp<T, B> {
+    Atom(B),
     Pair(T, T),
 }
 
 pub trait Allocator {
     type Ptr: Clone;
+    type AtomBuf: Clone;
 
-    fn new_atom(&self, v: &[u8]) -> Self::Ptr;
-    fn new_pair(&self, first: Self::Ptr, rest: Self::Ptr) -> Self::Ptr;
-    fn sexp<'a: 'c, 'b: 'c, 'c>(&'a self, node: &'b Self::Ptr) -> SExp<'c, Self::Ptr>;
+    fn new_atom(&mut self, v: &[u8]) -> Self::Ptr;
+    fn new_pair(&mut self, first: Self::Ptr, rest: Self::Ptr) -> Self::Ptr;
+
+    // The lifetime here is a bit special because IntAllocator and ArcAllocator
+    // have slightly different requirements. With IntAllocator, all buffers are
+    // owned by the allocator, with ArcAllocator all buffers have shared
+    // ownership by ArcAllocator::Ptr objects. So the returned buffer here
+    // depends on both
+    fn atom<'a>(&'a self, node: &'a Self::Ptr) -> &'a [u8];
+    fn buf<'a>(&'a self, node: &'a Self::AtomBuf) -> &'a [u8];
```

Comment:
Maybe we can just ensure that `Self::AtomBuf : Into<'a &[u8]>` ie. is a type that supports `Into<'a &[u8]>`, and then we don't need this extra `buf<'a>` step. The type could simply keep a pointer to the allocator (if it needs it, as `IntAllocator` does), then you just go `.into()` to turn it into a `&'a [u8]`.

Reply:
I guess the risk here is that now `sexp` might return a thing that keeps a reference to `'a`. But I think you've already done all the untangling work to make this possible.

Reply:
it cannot keep a pointer to the allocator. rust wouldn't allow creating any new atoms or pairs on the allocator while this object exists in that case. since you can't have aliases to mutable references

Reply:
Are you sure? Since `buf` returns something `'a`, rust treats that pointer as clinging to a non-mutable lifetime of the allocator anyway. You either do something with it or discard it, but in any case it has to be gone before you call `new_pair` or `new_atom` since it has lifetime `'a`.

Reply:
I'm pretty sure your untangling has made this possible. I really should try hacking the code though to prove it.

Reply:
rust lets that variable exist in the same basic block, and you can still use the aliasing mutable reference as long as that variable (that has the lifetime dependency) is not used later, in that path of the code

Reply:
I tried changing `AtomBuf` to `(&Allocator, u32)`, but it needs a lifetime, and generic types like `type AtomBuf<'a>` are not a stable feature at the moment.

```
error[E0658]: generic associated types are unstable
  --> src/int_allocator.rs:37:5
   |
37 |     type AtomBuf<'a> = (&'a Self, u32);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #44265 <https://github.com/rust-lang/rust/issues/44265> for more information
```